fix(compiler): transform pseudo selectors correctly for the encapsulated view

GeorgySerga opened the pull request:
fix scoping and transforming logic of the `shimCssText` for the components with encapsulated view:
- add support for pseudo selector functions
- apply content scoping for inner selectors of `:is()` and `:where()`
- allow multiple comma separated selectors inside pseudo selectors

Fixes #45686

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
Pseudo selector functions aren't parsed/transformed correctly at the moment
The provided examples inside the issue are added as unit tests as well as other test cases.
Issue Number: #45686 


## What is the new behavior?
Specified in unit tests.
- selectors are parsed by comma if comma is not inside the parenthesis, for example `:where(.one, .two)` stays as a single selector for scoping, previously it was split into two parts `:where(.one` and `.two)`.
- selector parts aren't broken into pieces by combinators if they are within parenthesis, for example `:where(.one > .two)` isn't split
- selectors within `:is()` and `:where()` are being scoped internally, that is especially important for `:where()` since this selectors doesn't increase specificity, allowing overloading theme specific values not risking order related issues.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

GeorgySerga committed:
fix(compiler): transform pseudo selectors correctly for the encapsulated view

fix scoping and transforming logic of the `shimCssText` for the
components with encapsulated view:
- add support for pseudo selector functions
- apply content scoping for inner selectors of `:is()` and `:where()`
- allow multiple comma separated selectors inside pseudo selectors

Fixes #45686

GeorgySerga committed:
fix(compiler): allow combinators inside pseudo selectors

allow css combinators within pseudo selector functions, parsing those
correctly. Similarly to previous version, don't break selectors
into part if combinators are within parenthesis, for example
`:where(.one > .two)`

GeorgySerga committed:
fix(compiler): fix comment typo

fix spelling in the comment

crisbeto commented:
Running a TGP since encapsulation changes tend to be break-ey.

GeorgySerga committed:
fix(compiler): add multiple :host and nested selectors support

add support for nested and deeply nested (up to three levels) selectors,
parse multiple :host selectors, scope selectors within pseudo functions

GeorgySerga commented:
Thank you @crisbeto ! I patched my change with more tests and support for more complex selectors. The internal presubmit test, which failed previously, passed with new changes.

crisbeto commented:
[I re-ran the TGP](http://test/OCL:675872382:BASE:677715311:1727086855961:da7abfdc) with the latest code, but there are still some breakages.

GeorgySerga committed:
fix(compiler): preserve attributes attached to :host selector

keep attributes used to scope :host selectors

GeorgySerga committed:
fix(compiler): fix parsing of the :host-context with pseudo selectors

fix regexp which is used to test for host inside pseudo selectors

GeorgySerga committed:
fix(compiler): scope :host-context inside pseudo selectors, do not decrease specificity

parse constructions like `:where(:host-context(.foo))` correctly
revert logic which lead to decreased specificity if `:where` was applied
to another selector, for example `div` is transformed to `div[contenta]`
with specificity of (0,1,1) so `div:where(.foo)` should not decrease it
leading to `div[contenta]:where(.foo)` with the same specificity (0,1,1)
instead of `div:where(.foo[contenta])` with specificity equal to (0,0,1)

crisbeto commented:
[Passing TGP](http://test/OCL:675872382:BASE:684040817:1728486456281:d7a9ae6f)

devversion commented:
This PR was merged into the repository by commit 46a6324c82a41b69c16a4c8c9f3fc52d1ecf6917.

The changes were merged into the following branches: main, 18.2.x
